subprojects/core/src/test/groovy/org/gradle/api/tasks/JavaExecTest.groovy

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.tasks
+
+import org.gradle.api.internal.AbstractTask
+import org.gradle.process.ExecResult
+import org.gradle.process.internal.ExecAction
+import org.gradle.process.internal.ExecException
+
+class JavaExecTest extends AbstractTaskTest {
```

Comment:
❌ These tests aren't passing. Either fix we them or write integration test instead of unit tests. The complicated issue with `JavaExec` is it's using more services than `Exec` so mocking may not be the right choice for this. The following patch was required for some basic failures this PR introduce:
```
$ g diff --patch
diff --git a/subprojects/core/src/main/java/org/gradle/api/tasks/JavaExec.java b/subprojects/core/src/main/java/org/gradle/api/tasks/JavaExec.java
index 7ae8653707e..07f9a4484cd 100644
--- a/subprojects/core/src/main/java/org/gradle/api/tasks/JavaExec.java
+++ b/subprojects/core/src/main/java/org/gradle/api/tasks/JavaExec.java
@@ -99,7 +99,7 @@
  * </pre>
  */
 public class JavaExec extends ConventionTask implements JavaExecSpec {
-    private final JavaExecAction javaExecHandleBuilder;
+    private JavaExecAction javaExecHandleBuilder;
     private ExecResult execResult;
 
     public JavaExec() {
@@ -688,4 +688,8 @@ public boolean isIgnoreExitValue() {
     public ExecResult getExecResult() {
         return execResult;
     }
+
+    void setExecAction(JavaExecAction execAction) {
+        this.javaExecHandleBuilder = execAction;
+    }
 }
diff --git a/subprojects/core/src/test/groovy/org/gradle/api/tasks/JavaExecTest.groovy b/subprojects/core/src/test/groovy/org/gradle/api/tasks/JavaExecTest.groovy
index d85bf3bb351..5958c1ae5a5 100644
--- a/subprojects/core/src/test/groovy/org/gradle/api/tasks/JavaExecTest.groovy
+++ b/subprojects/core/src/test/groovy/org/gradle/api/tasks/JavaExecTest.groovy
@@ -17,12 +17,12 @@ package org.gradle.api.tasks
 
 import org.gradle.api.internal.AbstractTask
 import org.gradle.process.ExecResult
-import org.gradle.process.internal.ExecAction
 import org.gradle.process.internal.ExecException
+import org.gradle.process.internal.JavaExecAction
 
 class JavaExecTest extends AbstractTaskTest {
     JavaExec execTask
-    def execAction = Mock(ExecAction) {
+    def execAction = Mock(JavaExecAction) {
         getArgumentProviders() >> []
     }
 ```

I did toy around adding integration tests which may be the proper way to test `JavaExec` as well as `Exec` using this patch:

```
$  g show 468cad3114199e720dde94f0d619dc7458d2ffcf --patch
commit 468cad3114199e720dde94f0d619dc7458d2ffcf (HEAD -> lacasseio/pr-10763-integ-test)
Author: Daniel Lacasse <daniel@gradle.com>
Date:   Sat Nov 9 10:44:55 2019 +0300

    WIP

diff --git a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/AbstractExecTaskIntegrationTest.groovy b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/AbstractExecTaskIntegrationTest.groovy
new file mode 100644
index 00000000000..2f93883394e
--- /dev/null
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/AbstractExecTaskIntegrationTest.groovy
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+
+abstract class AbstractExecTaskIntegrationTest extends AbstractIntegrationSpec {
+    protected abstract void makeExecProject()
+    protected abstract void writeSucceedingExec()
+    protected abstract void writeFailingExec()
+
+    protected String getTaskUnderTestDsl() {
+        return "tasks.${taskNameUnderTest}"
+    }
+
+    protected abstract String getTaskNameUnderTest()
+
+    def "returns null ExecResult when task haven't executed"() {
+        makeExecProject()
+        writeSucceedingExec()
+        buildFile << """
+            task verify {
+                doLast {
+                    assert ${taskUnderTestDsl}.execResult == null
+                }
+            }
+        """
+
+        when:
+        succeeds('verify')
+
+        then:
+        result.assertTasksExecutedAndNotSkipped(':verify')
+    }
+
+    def "returns ExecResult when is executed"() {
+        makeExecProject()
+        writeSucceedingExec()
+        buildFile << """
+            task verify {
+                dependsOn(${taskNameUnderTest})
+                doLast {
+                    assert ${taskUnderTestDsl}.execResult != null
+                    assert ${taskUnderTestDsl}.execResult.exitValue == 0
+                }
+            }
+        """
+
+        when:
+        succeeds('verify')
+
+        then:
+        result.assertTasksExecutedAndNotSkipped(':compileJava', ":$taskNameUnderTest", ':verify')
+    }
+
+    def "execute with non-zero exit value and ignore exit value should not throw exception"() {
+        makeExecProject()
+        writeFailingExec()
+        buildFile << """
+            ${taskUnderTestDsl}.ignoreExitValue = true
+            
+            task verify {
+                dependsOn(${taskNameUnderTest})
+                doLast {
+                    assert ${taskUnderTestDsl}.execResult != null
+                    assert ${taskUnderTestDsl}.execResult.exitValue == 42
+                }
+            }
+        """
+        when:
+        succeeds('verify')
+
+        then:
+        result.assertTasksExecutedAndNotSkipped(':compileJava', ":$taskNameUnderTest", ':verify')
+    }
+}
diff --git a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/ExecTaskIntegrationTest.groovy b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/ExecTaskIntegrationTest.groovy
new file mode 100644
index 00000000000..43e39dd3efd
--- /dev/null
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/ExecTaskIntegrationTest.groovy
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks
+
+import org.gradle.internal.jvm.Jvm
+import org.gradle.test.fixtures.file.TestFile
+
+class ExecTaskIntegrationTest extends AbstractExecTaskIntegrationTest {
+    TestFile mainJavaFile
+
+    @Override
+    protected void makeExecProject() {
+        buildFile.text = """
+            apply plugin: "java"
+
+            task run(type: Exec) {
+                dependsOn(compileJava)
+                executable = ${Jvm.canonicalName}.current().javaExecutable
+                args '-cp', project.layout.files(compileJava).asPath, 'driver.Driver', "1"
+            }
+        """
+    }
+
+    @Override
+    protected void writeSucceedingExec() {
+        mainJavaFile = file('src/main/java/Driver.java')
+        file("src/main/java/Driver.java").text = mainClass("""
+            try {
+                FileWriter out = new FileWriter("out.txt");
+                for (String arg: args) {
+                    out.write(arg);
+                    out.write("\\n");
+                }
+                out.close();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        """)
+    }
+
+    @Override
+    protected void writeFailingExec() {
+        mainJavaFile = file('src/main/java/Driver.java')
+        file("src/main/java/Driver.java").text = mainClass("""
+            System.exit(42);
+        """)
+    }
+
+    @Override
+    protected String getTaskNameUnderTest() {
+        return "run"
+    }
+
+    private static String mainClass(String body) {
+        return """
+            package driver;
+
+            import java.io.*;
+            import java.lang.System;
+
+            public class Driver {
+                public static void main(String[] args) {
+                ${body}
+                }
+            }
+        """
+    }
+}
diff --git a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/JavaExecTaskIntegrationTest.groovy b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/JavaExecTaskIntegrationTest.groovy
new file mode 100644
index 00000000000..086a9ca43a0
--- /dev/null
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/JavaExecTaskIntegrationTest.groovy
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks
+
+import org.gradle.test.fixtures.file.TestFile
+
+class JavaExecTaskIntegrationTest extends AbstractExecTaskIntegrationTest {
+    TestFile mainJavaFile
+
+    @Override
+    protected void makeExecProject() {
+        buildFile.text = """
+            apply plugin: "java"
+
+            task run(type: JavaExec) {
+                classpath = project.layout.files(compileJava)
+                main "driver.Driver"
+                args "1"
+            }
+        """
+    }
+
+    @Override
+    protected void writeSucceedingExec() {
+        mainJavaFile = file('src/main/java/Driver.java')
+        file("src/main/java/Driver.java").text = mainClass("""
+            try {
+                FileWriter out = new FileWriter("out.txt");
+                for (String arg: args) {
+                    out.write(arg);
+                    out.write("\\n");
+                }
+                out.close();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        """)
+    }
+
+    @Override
+    protected void writeFailingExec() {
+        mainJavaFile = file('src/main/java/Driver.java')
+        file("src/main/java/Driver.java").text = mainClass("""
+            System.exit(42);
+        """)
+    }
+
+    @Override
+    protected String getTaskNameUnderTest() {
+        return "run"
+    }
+
+    private static String mainClass(String body) {
+        return """
+            package driver;
+
+            import java.io.*;
+            import java.lang.System;
+
+            public class Driver {
+                public static void main(String[] args) {
+                ${body}
+                }
+            }
+        """
+    }
+}
```